src/qibocal/calibrations/niGSC/basics/experiment.py

```
@@ -0,0 +1,223 @@
+        ), """
+            given circuit factory has wrong type {}, must be qibo NoiseModel | None .""".format(
+            type(noise_model)
+        )
+        self.__noise_model = noise_model
+        self.name = "Abstract"
+
+    @property
+    def noise_model(self):
+        return self.__noise_model
+
+    @property
+    def dataframe(self) -> pd.DataFrame:
+        return pd.DataFrame(self.data)
+
+    @classmethod
+    def load(cls, path: str) -> Experiment:
+        """Creates an experiment object with data and if possible with circuits.
+
+        Args:
+            path (str): The directory from where the object should be restored.
+
+        Returns:
+            Experiment: The object with data (and circuitfactory).
+        """
+        datapath = f"{path}experiment_data.pkl"
+        circuitspath = f"{path}circuits.pkl"
+        if isfile(datapath):
+            with open(datapath, "rb") as f:
+                data = pickle.load(f)
+            if isinstance(data, pd.DataFrame):  # pragma: no cover
+                data = data.to_dict("records")
+            nshots = len(data[0]["samples"])
+        else:
+            data, nshots = None, None
+        if isfile(circuitspath):
+            with open(circuitspath, "rb") as f:
+                circuitfactory = pickle.load(f)
+        else:
+            circuitfactory = None
+        # Initiate an instance of the experiment class.
+        obj = cls(circuitfactory, data=data, nshots=nshots)
+        return obj
+
+    def save(self, path: str | None = None) -> str:
+        """Creates a path if None given and pickles relevant data from ``self.data``
+        and if ``self.circuitfactory`` is a list that one too.
+
+        Returns:
+            (str): The path of stored experiment.
+        """
+
+        # Check if path to store is given, if not create one. If yes check if the last character
+        # is a /, if not add it.
+        if path is None:
+            self.path = experiment_directory("rb")
```

I think that you can simply raise an error if `path` is `None`. You can also remove the function `experiment_directory` if it is not used in other parts of the code. In this way we can avoid code duplication given that the `ActionBuilder` already does a similar thing.

---

I want the saving and loading methods to be independent from ``qq``.

---

I understand your point. However, I still think that it will be sufficient to specify the path given by the user without having to serialize the output folder given that this functionality is provided by qq. Let's say also what the other reviewers think.

---

If I understand correctly @andrea-pasquale means the `_generate_output_folder` function of the `ActionBuilder`. I would agree to reuse this method here to avoid code repetition, however if you do this, I would suggest to move this method outside the `ActionBuilder` (it is a `@staticmethod` anyway) and remove the underscore so that it can be used in other parts.

> I think that you can simply raise an error if `path` is `None`.

If you go with this approach, I think it is simpler to just remove the default `None` argument and make `path` required, instead of raising an error if `None`.